src/mumble/pluginAPI/PluginComponents.h

```diff
@@ -0,0 +1,232 @@
+/**
+ * This header file contains definitions of types and other components used in Mumble's plugin system
+ */
+
+#ifndef MUMBLE_PLUGINCOMPONENT_H_
+#define MUMBLE_PLUGINCOMPONENT_H_
+
+#include <stdint.h>
+#include <stddef.h>
+
+#define STATUS_OK EC_OK
+
+enum PluginFeature {
+	/// None of the below
+	FEATURE_NONE = 0,
+	/// The plugin provides positional data from a game
+	FEATURE_POSITIONAL = 2 << 0,
```

Comment:
The bitshifting makes it obvious that these are bitflags.

Reply:
I just realized that `2 << 0` isn't even 1 as I intended :man_facepalming: 
I'll change it to 1 then. There is no way to represent a 1 with a left-shift operator, is there?

Reply:
`1 << 0` is a one WITH a left shift operator in it

Reply:
True - I settled on `2 >> 1` instead though as I think this holds up the spirit of showing that these flags are all powers of 2

Reply:
> that these flags are all powers of 2
except for one?
But then you do a right shift just to get one, and a left shift for all the others?
why not just stay with one style?

Reply:
> except for 1?

1 is also a power of two: `2^0 = 1`

> why not just stay with one style?

The style I am trying to stick to is that all bitflags are derived from the number 2 (except the 0 value which is special).

Reply:
except 0, and except 1.
Currently you have 3 different styles for the same thing

```
0,
2 >> 1,
2 << 0
2 << 1
2 << 2
```
I would prefer
```
0
1 << 0
1 << 1
1 << 2
1 << 3
```

`2 << 1` tells you the third bit is set? Weird, it says 1 though..
`1 << 2` tells you the third bit is set. Arrays start from 0, so a programmer would understand that better.

Reply:
Oh yeah that's better - totally missed that we don't have to start from 2 for this (guess my head was too wrapped around the mathematical description xD).
I'll change it accordingly

Reply:
Maybe you should #define some constants instead?

Reply:
You mean something like this?
```
#define FEATURE_NONE 0
#define FEATURE_POSITIONAL 1
#define FEATURE_X 2
#define FEATURE_Y 4
#define FEATURE_Z 8
```

What's the benefit of that?

Reply:
Yeah, that way it is kind of self documenting what each bit means, and you avoid mistakes like putting "<<" the wrong way or whatnot. Might not make a huge difference here, but will hopefully make it less likely that the bits gets confused in plugins later on, and increases the chances that a typo results in a compile error instead of just weird behavior. Just my opinion :)

Reply:
> that way it is kind of self documenting what each bit means

We have that with the enum as well. In fact for the end user the usage of the defines is just the same as you'd use the enum constants.

> you avoid mistakes like putting "<<" the wrong way or whatnot. Might not make a huge difference here, but will hopefully make it less likely that the bits gets confused in plugins later on

The end-user won't use the shift-operator at all. In a plugin you'd simply do something like
```
return FEATURE_POSITIONAL;
```
and don't actually care about what particular value that constant has (here you can also see that the usage is the same as if the constant had been #defined).
Plus the current version uses only the left-shift operator so no confusion should come in there when adding new constants to the enum :)

> typo results in a compile error instead of just weird behavior

It does that already


Overall I don't really see any advantage of using #defines instead of the enum so I'll leave it at that :)

Reply:
Okay, that is good to hear. It was only a suggestion, because at some point, some plugin will probably want to compute the bit pattern themselves. I just think bit patterns should often be `#defined` very early when developing APIs so that you don't have to go back and add them later in thousands of places.

Reply:
And then you could do stuff like this:
```c
#define FEATURE_AXIS_X       1
#define FEATURE_AXIS_Y       2
#define FEATURE_AXIS_Z       4
#define FEATURE_ENVIRONMENTAL_AWARE      8
#define FEATURE_2D      (FEATURE_AXIS_X | FEATURE_AXIS_Y)
#define FEATURE_3D      (FEATURE_AXIS_X | FEATURE_AXIS_Y | FEATURE_AXIS_Z)
```

Then a 2D game could `return FEATURE_2D | FEATURE_ENVIRONMENTAL_AWARE;` if it could be that game want to indicate X,Y position and indicate if the player is, say, under water. Just thinking future proofing the API. May not be relevant in this case though, I am just brainstorming here :) Sorry if this is just noise with regards to this MR

Reply:
> I just think bit patterns should often be #defined very early when developing APIs so that you don't have to go back and add them later in thousands of places.

In principle I agree but in this case it doesn't really take much to add another feature later on. Just add an entry to the enum and implement the new flag in the backend of the plugin framework. Done.


The different axes aren't necessary. If the supported game is 2D then the plugin still can return a 3D array with z=0. Everything else would also require the backend for the positional audio to be changed (quite a bit).
I do like the environmental aware thing, but it only makes sense to have such a flag if the backend actually uses that kind of information (which atm it doesn't). Thus I think it is enough to add that flag once the corresponding feature is supported by the backend :)

> May not be relevant in this case though, I am just brainstorming here :) Sorry if this is just noise with regards to this MR

Nothing to be sorry for - all thoughts and ideas are very welcome ;)